common/src/primitives/merkle/tree.rs

```diff
@@ -0,0 +1,560 @@
+// Copyright (c) 2021-2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{collections::BTreeSet, num::NonZeroUsize};
+
+use crypto::hash::StreamHasher;
+use itertools::Itertools;
+
+use crate::primitives::{
+    id::{default_hash, DefaultHashAlgoStream},
+    H256,
+};
+
+use super::{MerkleTreeAccessError, MerkleTreeFormError, MerkleTreeProofExtractionError};
+
+pub enum AdjacentLeavesIndices {
+    Alone(usize),
+    Together(usize, usize),
+}
+
+/// Merkle tree in the form of a vector, where the bottom leaves are the based, and the root is
+/// the last element.
+#[derive(Debug)]
+pub struct MerkleTree {
+    tree: Vec<H256>,
+}
+
+fn next_pow2(n: usize) -> usize {
+    if n == 0 {
+        return 1;
+    }
```

Comment:
Hmm looks like a function for doing this has been stabilised since it was first implemented

https://doc.rust-lang.org/std/primitive.usize.html#method.checked_next_power_of_two

Reply:
Nice!